<a href="https://colab.research.google.com/github/Tirta2241720045/Machine_Learning/blob/main/Minggu5_2241720045_TirtaNBP_Tugas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tugas 1

1. Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv.

In [ ]:
import pandas as pd # Import library pandas
from sklearn.svm import SVC # Import library SVM dari sklearn
from sklearn.model_selection import train_test_split    # Import library train_test_split dari sklearn
from sklearn.metrics import accuracy_score  # Import library accuracy_score dari sklearn
from sklearn.preprocessing import StandardScaler, LabelEncoder  # Import library StandardScaler, LabelEncoder dari sklearn

# Muat dataset suara
df = pd.read_csv('/content/drive/MyDrive/dataset/voice.csv')    # Load dataset suara

# Pisahkan fitur dan label
X = df.drop(columns=['label'])  # Fitur dataset suara
y = df['label'] # Label dataset suara

# Skala fitur
X = StandardScaler().fit_transform(X)   # Skala fitur dataset suara

# Encode label
y = LabelEncoder().fit_transform(y) # Encode label dataset suara

# Bagi data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)   # Bagi dataset suara menjadi data latih dan data uji

# Latih model SVM
svm_model = SVC()   # Inisialisasi model SVM
svm_model.fit(X_train, y_train) # Latih model SVM

# Lakukan prediksi dan evaluasi akurasi
accuracy = accuracy_score(y_test, svm_model.predict(X_test))    # Evaluasi akurasi model SVM
print(f"Prediction accuracy: {accuracy:.4f}")   # Print akurasi model SVM


Prediction accuracy: 0.9795


### Hasil
Akurasi model:
```
Prediction accuracy: 0.9795
```

### Kesimpulan
Model SVM menunjukkan akurasi tinggi, menandakan efektivitasnya dalam klasifikasi suara.

2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
  1. Menggunakan data spam.csv
  2. Fitur CountVectorizer dengan mengaktifkan stop_words
  3. Evaluasi hasilnya

In [ ]:
import pandas as pd # Import library pandas
from sklearn.model_selection import train_test_split    # Import library train_test_split dari sklearn
from sklearn.feature_extraction.text import CountVectorizer   # Import library CountVectorizer dari sklearn
from sklearn.naive_bayes import MultinomialNB   # Import library MultinomialNB dari sklearn
from sklearn.metrics import accuracy_score, classification_report   # Import library accuracy_score, classification_report dari sklearn

# Muat dataset spam
df = pd.read_csv('/content/drive/MyDrive/dataset/spam.csv', encoding='latin-1').iloc[:, :2]   # Load dataset spam
df.columns = ['Labels', 'SMS']  # Rename kolom dataset spam

# Pisahkan fitur dan label
X = df['SMS']   # Fitur dataset spam
y = df['Labels']    # Label dataset spam

# Bagi data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)   # Bagi dataset spam menjadi data latih dan data uji

# Inisiasi CountVectorizer dengan stop_words
bow = CountVectorizer(stop_words='english')   # Inisialisasi CountVectorizer dengan stop_words
X_train_bow = bow.fit_transform(X_train)    # Transformasi data latih
X_test_bow = bow.transform(X_test)  # Transformasi data uji

# Fit model Naive Bayes
mnb = MultinomialNB().fit(X_train_bow, y_train)   # Inisialisasi model Naive Bayes

# Prediksi dan evaluasi data pelatihan
y_pred_train = mnb.predict(X_train_bow)   # Prediksi data latih
acc_train = accuracy_score(y_train, y_pred_train)   # Evaluasi akurasi data latih

# Prediksi dan evaluasi data pengujian
y_pred_test = mnb.predict(X_test_bow)   # Prediksi data uji
acc_test = accuracy_score(y_test, y_pred_test)  # Evaluasi akurasi data uji

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train:.4f}')  # Print akurasi data latih
print(f'Hasil akurasi data test: {acc_test:.4f}')   # Print akurasi data uji
print('Report Klasifikasi (Data Test):\n', classification_report(y_test, y_pred_test))  # Print report klasifikasi data uji

Hasil akurasi data train: 0.9946
Hasil akurasi data test: 0.9830
Report Klasifikasi (Data Test):
               precision    recall  f1-score   support

         ham       0.98      1.00      0.99       954
        spam       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



### Hasil

Dari hasil running program, berikut adalah output yang diperoleh:

- **Hasil Akurasi Data Train**:
  ```
  Hasil akurasi data train: 0.9946
  ```

- **Hasil Akurasi Data Test**:
  ```
  Hasil akurasi data test: 0.9830
  ```

- **Report Klasifikasi (Data Test)**:
  ```
                  precision    recall  f1-score   support

            ham       0.98      1.00      0.99       954
           spam       0.98      0.90      0.94       161

       accuracy                           0.98      1115
  ```

### Kesimpulan

1. **Akurasi**: Model mencapai akurasi 99.46% pada data train dan 98.30% pada data test, menunjukkan performa yang sangat baik dalam klasifikasi SMS.
2. **Kinerja**: Sangat efektif dalam mendeteksi "ham" (Recall 1.00) dan cukup baik untuk "spam" (Recall 0.90).
3. **Keseimbangan**: Meskipun model baik dalam mendeteksi "ham", ada ruang untuk meningkatkan deteksi "spam" agar lebih seimbang.


3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
  1. Menggunakan data spam.csv
  2. Fitur TF-IDF dengan mengaktifkan stop_words
  3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
  4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [ ]:
import pandas as pd # Import library pandas
from sklearn.model_selection import train_test_split    # Import library train_test_split dari sklearn
from sklearn.feature_extraction.text import TfidfVectorizer   # Import library TfidfVectorizer dari sklearn
from sklearn.naive_bayes import MultinomialNB   # Import library MultinomialNB dari sklearn
from sklearn.metrics import accuracy_score, classification_report   # Import library accuracy_score, classification_report dari sklearn

# Muat dataset spam
df = pd.read_csv('/content/drive/MyDrive/dataset/spam.csv', encoding='latin-1').iloc[:, :2]  # Load dataset spam
df.columns = ['Labels', 'SMS']  # Rename kolom dataset spam

# Pisahkan fitur dan label
X = df['SMS']   # Fitur dataset spam
y = df['Labels']    # Label dataset spam

# Bagi data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)   # Bagi dataset spam menjadi data latih dan data uji

# Inisiasi TfidfVectorizer dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')   # Inisialisasi TfidfVectorizer dengan stop_words
X_train_tfidf = tfidf.fit_transform(X_train)    # Transformasi data latih
X_test_tfidf = tfidf.transform(X_test)  # Transformasi data uji

# Fit model Multinomial Naive Bayes
mnb = MultinomialNB().fit(X_train_tfidf, y_train)   # Inisialisasi model Multinomial Naive Bayes

# Prediksi dan evaluasi data pelatihan
y_pred_train = mnb.predict(X_train_tfidf)   # Prediksi data latih
acc_train = accuracy_score(y_train, y_pred_train)   # Evaluasi akurasi data latih

# Prediksi dan evaluasi data pengujian
y_pred_test = mnb.predict(X_test_tfidf)  # Prediksi data uji
acc_test = accuracy_score(y_test, y_pred_test)  # Evaluasi akurasi data uji

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train:.4f}') # Print akurasi data latih
print(f'Hasil akurasi data test: {acc_test:.4f}')   # Print akurasi data uji
print('Report Klasifikasi (Data Test):\n', classification_report(y_test, y_pred_test))  # Print report klasifikasi data uji

Hasil akurasi data train: 0.9843
Hasil akurasi data test: 0.9605
Report Klasifikasi (Data Test):
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       954
        spam       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



### Hasil

- **Hasil Akurasi Data Train**: 0.9843
- **Hasil Akurasi Data Test**: 0.9605

- **Report Klasifikasi (Data Test)**:
  ```
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       954
        spam       1.00      0.73      0.84       161

    accuracy                           0.96      1115
  ```

### Kesimpulan

1. **Akurasi**: Model mencapai akurasi 98.43% pada data train dan 96.05% pada data test.
2. **Kinerja**: Sangat efektif mendeteksi "ham", tetapi kurang optimal untuk "spam" (Recall 0.73).
3. **Rekomendasi**: Perbaikan diperlukan untuk mendeteksi spam lebih baik, mungkin dengan teknik penyeimbangan data.

# Tugas 2
Buatlah model klasifikasi dengan menggunakan SVM berdasarkan data pada Percobaan 5 dengan menggunakan **fitur histogram.**

### 1. **Library Import**

In [ ]:
from pathlib import Path    # Import library Path dari pathlib
import matplotlib.pyplot as plt   # Import library matplotlib.pyplot
import cv2  # Import library cv2
import numpy as np  # Import library numpy
import pandas as pd # Import library pandas
from sklearn.svm import SVC # Import library SVM dari sklearn
from sklearn.metrics import accuracy_score  # Import library accuracy_score dari sklearn

- **Langkah**: Mengimpor library yang dibutuhkan.
- **Fungsi**:
  - `Path` untuk navigasi file sistem.
  - `cv2` (OpenCV) untuk operasi gambar seperti membaca gambar, konversi warna, dan menghitung histogram.
  - `numpy` untuk operasi array.
  - `pandas` untuk penyimpanan data dalam bentuk DataFrame.
  - `SVC` dari `sklearn` untuk membuat model SVM (Support Vector Machine).
  - `accuracy_score` dari `sklearn` untuk menghitung akurasi model.

### 2. **Fungsi Load Dataset**


In [ ]:
def load_dataset(img_dir):  # Fungsi untuk memuat dataset
    p = Path(img_dir)   # Path dataset
    dirs = p.glob('*')  # List direktori dataset

    img_list = []   # List dataset

    for dir in dirs:    # Looping setiap direktori
        label = str(dir).split('/')[-1] # Label dataset
        for file in dir.glob('*.jpg'):  # Looping setiap file gambar
            img = cv2.imread(str(file)) # Baca gambar

            if img is not None: # Jika gambar terbaca
                img_list.append((img, label))   # Tambahkan gambar dan label ke list

    return img_list # Return list dataset

- **Langkah**: Memuat dataset gambar.
- **Fungsi**:
  - Membaca gambar dari folder `train_dir` dan `test_dir`.
  - Setiap gambar disimpan sebagai tuple berisi gambar itu sendiri (array) dan label (nama folder di mana gambar itu berada).
- **Alur Kerja**: Fungsi ini mengembalikan daftar gambar dan label, yang akan digunakan untuk ekstraksi fitur.

### 3. **Load Training dan Test Data**

In [ ]:
train_dir = "/content/drive/MyDrive/dataset/images/training"    # Direktori data latih
test_dir = "/content/drive/MyDrive/dataset/images/test" # Direktori data uji

train_img = load_dataset(train_dir) # Memuat dataset latih
test_img = load_dataset(test_dir)   # Memuat dataset uji

- **Langkah**: Memanggil fungsi `load_dataset` untuk memuat gambar dari folder `training` dan `test`.
- **Fungsi**:
  - Gambar dari folder `train_dir` dimuat ke dalam `train_img`, dan dari `test_dir` dimuat ke dalam `test_img`.

### 4. **Fungsi Ekstraksi Fitur Histogram**

In [ ]:
def extract_histogram_features(img_list, bins=32):  # Fungsi untuk mengekstrak fitur histogram
    hist_features = []  # List fitur histogram
    labels = [] # List label

    for img, label in img_list: # Looping setiap gambar
        gray_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)    # Konversi gambar ke grayscale

        hist = cv2.calcHist([gray_img], [0], None, [bins], [0, 256])    # Hitung histogram gambar
        hist = cv2.normalize(hist, hist).flatten()  # Normalisasi histogram

        hist_features.append(hist)  # Tambahkan histogram ke list
        labels.append(label)    # Tambahkan label ke list

    df = pd.DataFrame(hist_features)    # Konversi list fitur histogram ke dataframe
    df['LABELS'] = labels   # Tambahkan kolom label ke dataframe

    return df   # Return dataframe

- **Langkah**: Ekstraksi fitur histogram.
- **Fungsi**:
  - Gambar diubah menjadi grayscale.
  - Histogram dihitung menggunakan fungsi `cv2.calcHist`, dan hasilnya dinormalisasi serta disimpan dalam bentuk array.
  - Fitur histogram dan label dimasukkan ke dalam DataFrame untuk analisis lebih lanjut.
- **Alur Kerja**: Fitur histogram dari setiap gambar digunakan sebagai input untuk pelatihan model.

### 5. **Ekstraksi Fitur Histogram untuk Data Training dan Test**


In [ ]:
train_hist_img = extract_histogram_features(train_img)  # Ekstrak fitur histogram data latih
test_hist_img = extract_histogram_features(test_img)    # Ekstrak fitur histogram data uji

- **Langkah**: Memanggil fungsi `extract_histogram_features` untuk menghitung histogram dari gambar training dan test.
- **Fungsi**: Hasilnya disimpan dalam DataFrame `train_hist_img` dan `test_hist_img`, di mana setiap kolom menyimpan histogram, dan kolom `LABELS` menyimpan label.

### 6. **Memisahkan Data dan Label**

In [ ]:
X_train = train_hist_img.drop(columns=['LABELS']).values    # Fitur data latih
y_train = train_hist_img['LABELS']  # Label data latih

X_test = test_hist_img.drop(columns=['LABELS']).values  # Fitur data uji
y_test = test_hist_img['LABELS']    # Label data uji

- **Langkah**: Memisahkan fitur dan label.
- **Fungsi**:
  - `X_train` dan `X_test` adalah data fitur histogram, sedangkan `y_train` dan `y_test` adalah label yang sesuai.
- **Alur Kerja**: Data fitur digunakan untuk melatih model SVM, dan label digunakan untuk evaluasi.

### 7. **Pelatihan Model SVM**


In [ ]:
model = SVC()   # Inisialisasi model SVM
model.fit(X_train, y_train) # Latih model SVM

SVC()

- **Langkah**: Membuat dan melatih model SVM.
- **Fungsi**:
  - Model SVM dibuat menggunakan `SVC` dan dilatih dengan data training (`X_train`, `y_train`).
- **Alur Kerja**: Model belajar pola dari histogram untuk membedakan antara gambar day/night.


### 8. **Evaluasi Model**

In [ ]:
y_train_pred = model.predict(X_train)   # Prediksi data latih
train_accuracy = accuracy_score(y_train, y_train_pred)  # Evaluasi akurasi data latih

y_test_pred = model.predict(X_test) # Prediksi data uji
test_accuracy = accuracy_score(y_test, y_test_pred)   # Evaluasi akurasi data uji

- **Langkah**: Membuat prediksi pada data training dan test.
- **Fungsi**:
  - `predict` digunakan untuk membuat prediksi pada data training dan test.
  - Akurasi dihitung menggunakan `accuracy_score`.
- **Alur Kerja**: Model dievaluasi dengan menghitung akurasi pada data training dan test.

### 9. **Hasil Evaluasi**

In [ ]:
print(f'Accuracy on training data: {train_accuracy}')   # Print akurasi data latih
print(f'Accuracy on test data: {test_accuracy}')    # Print akurasi data uji

Accuracy on training data: 1.0
Accuracy on test data: 0.95625


- **Langkah**: Menampilkan akurasi model.
- **Fungsi**:
  - Menampilkan akurasi model pada data training dan test.
- **Alur Kerja**: Hasil akhir memberikan gambaran seberapa baik model SVM dalam mengklasifikasikan gambar berdasarkan histogram.

### **Kesimpulan dari Hasil**
- Akurasi pada **training data** menunjukkan seberapa baik model mengenali pola pada data yang telah digunakan untuk pelatihan.
- Akurasi pada **test data** menunjukkan seberapa baik model menggeneralisasi pola untuk data baru yang tidak pernah dilihat sebelumnya.

Apabila akurasi training dan test cukup tinggi, dapat disimpulkan bahwa fitur histogram cukup baik dalam membedakan gambar **day** dan **night**. Jika ada perbedaan signifikan antara akurasi training dan test, maka mungkin ada overfitting atau underfitting pada model.